enhancements/installer/single-node-installation-bootstrap-in-place.md

```diff
@@ -0,0 +1,256 @@
+1. The obvious additional node, in most cases there isn't additional Hardware/VM at the edge site.
+2. Requires external dependencies:
+   a. Load balancer (only for bootstrap phase)
+   b. Requires DNS (configured per installation)
+3. Cannot use Bare Metal IPI:
+   a. Adds irrelevant dependencies - vips, keepalived, mdns
+   b. Requires same L2 between bootstrap and the Single Node Openshift
+
+### Goals
+
+* This enhancement describes an approach for insallting Single Node OpenShift 
+  for production use.
+* Minimal changes to openshift installer, this implementation shouldn't effect existing deployment flows.
+* Installation should result a clean Single Node Openshift, without bootstrap leftovers.
+* Rely on a livecd, so it would be able to trigger the full installation using virtual media.
+* Self-Contained bootstrap, The bootstrap flow is expressed as a single Ignition config which is layered on top of a pristine RHCOS image. It's important that we limit our use of this pattern to just the Ignition configs - as opposed to including other assets directly in the ISO image. 
+
+### Non-Goals
+
+* This enhancement does not address similar insallation flow for 3 nodes cluster.
+* This enhancement does not address high-availability for single-node
+  deployments.
+* This enhancement does not address single-node-developer cluster-profile installation.
+
+## Proposal
+
+Execute the bootstrap flow on a liveCD.
+Enrich the master ignition with the control plane static pods and write this ignition.
+Write the updated master ignition + RHCOS to disk and reboot.
+Post reboot the node will complete the installation.
+
+### User Stories
+
+#### As a user, I can deploy OpenShift in a supported single-node configuration
+
+A user will be able to run the OpenShift installer to create a single-node
+ignition configuration.
+The user will boot RHCOS liveCD with this ignition to create a single-node deployment.
+
+### Implementation Details/Notes/Constraints
+
+The openshift installer `create ignition-configs` command will generate a `bootstrap-in-place.ign`
+file, when the number of replicas for the control plane (in the `install-config.yaml`) is `1`.
+
+The `bootstrap-in-place.ign` will be embedded into an RHCOS liveCD using coreos-install embed command.
+
+The user will boot a machine with this liveCD, the liveCD will start executing a similar flow as a bootstrap node in a regular installation:
+bootkube.sh will do all the rendering logic.
+The bootstrap static pods should be generated in a way that the control plane operators will be able to identify them and either continues in a controlled way for the next revision, or just keep them as the correct revision and reuse them.
+
+cluster-bootstrap will apply all the required manifests (under /opt/openshift/manifests/)
+
+Bootkube will get the master ignition from machine-config-server and
+enrich the master ignition with control plane static pods manifests and all required resources including etcd data.
```

Comment:
I think this is fine to start since we're not yet trying to handle in-place updates.  But just a note: having them in Ignition but not MachineConfig means they become "unmanaged state"; xref https://github.com/openshift/machine-config-operator/blob/master/docs/FAQ.md#q-if-i-change-something-manually-on-the-host-will-the-mco-revert-it

Reply:
I think unmanaged is OK since these files will be removed once the installation complete.